### PR TITLE
iOS: Enable views to be nested within <Text>

### PR DESCRIPTION
**Pull request by rigdern**
Previously, only Text and Image could be nested within Text. Now, any
view can be nested within Text. One restriction of this feature is
that developers must give inline views a width and a height via
the style prop.

Previously, inline Images were supported by using iOS's built-in support
for rendering images with an NSAttributedString via NSTextAttachment.
However, NSAttributedString doesn't support rendering arbitrary views.

This change adds support for nesting views within Text by creating one
NSTextAttachment per inline view. The NSTextAttachments act as placeholders.
They are set to be the size of the corresponding view. After the text is
laid out, we query the text system to find out where it has positioned each
NSTextAttachment. We then position the views to be at those locations.

This commit also contains a change in `RCTShadowText.m`
`_setParagraphStyleOnAttributedString:heightOfTallestSubview:`. It now only sets
`lineHeight`, `textAlign`, and `writingDirection` when they've actually changed. This
avoids unnecessarily dirtying the shadow tree which is important in some inline
view scenarios.

Here's a screen shot demoing an inline view and an inline image:

image

Adam Comella
Microsoft Corp.

### CODE REVIEW

**Comment by facebook-github-bot-9** [comment]
@rigdern updated the pull request.

File: Libraries/Image/Image.ios.js [hunk 1]
```diff
@@ -220,18 +216,11 @@ var Image = React.createClass({
     if (isNetwork && (tintColor || this.props.blurRadius)) {
       RawImage = RCTImageView;
     }
-
```
Comment (rigdern) [inline Libraries/Image/Image.ios.js]: How do I run the linter locally?

**Comment by facebook-github-bot-9** [comment]
@rigdern updated the pull request.

**Comment by nicklockwood** [comment]
@facebook-github-bot import

**Comment by facebook-github-bot-5** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1018155508272793/int_phab) to review.

**Comment by facebook-github-bot-8** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1018155508272793/int_phab) to review.

**Comment by nicklockwood** [comment]
This looks great!

I'm not sure the way that you're handling vertical alignment is correct though. It seems like you're basically aligning the views to the top of the line rect, but I would expect them to be aligned with the baseline of the text by default, which is how iOS normally positions text attachments, and how inline images were aligned on iOS previously.

Baseline alignment isn't always desirable, and we should probably consider ways to allow the developer to adjust the offset, as well as ensuring consistency with Android, but for now I'd like to ensure we remain consistent with how this worked on iOS previously.

**Comment by facebook-github-bot-7** [comment]
@rigdern updated the pull request.

**Comment by rigdern** [comment]
@nicklockwood I updated the PR to use 0.0 for the y coordinate like it did before my change. Hopefully the PR is good to go now.

Here are some questions/comments about future work:

If I wanted to support customizing the vertical alignment of inline views, what would you recommend? A new `style` on `View` that only works on inline views and is a no-op on other views (e.g. `inlineVerticalAlignment`with values like `"bottom"` and `"baseline"`)?

We're currently working on adding a similar feature to the Android implementation.

Adam Comella
Microsoft Corp.

**Comment by nicklockwood** [comment]
This looks good. I'll need to do some testing with our internal apps just to check that dropping support for automatically detecting image size hasn't broken anything.

> If I wanted to support customizing the vertical alignment of inline views, what would you recommend? A new style on View that only works on inline views and is a no-op on other views (e.g. inlineVerticalAlignmentwith values like "bottom" and "baseline")?

That's a good question. A simple solution would just be to respect the existing "top" and/or "marginTop" styles to adjust the view by a relative offset. That should work pretty-much out of the box on iOS since NSTextAttachment respects the bounds.origin and treats it as a relative offset (although we might want to flip the y value since it follows the Mac convention of y pointing up the screen instead of down IIRC).

Having a verticalAlignment property that follows the CSS spec would be nice as well, although personally I've never found it very useful in practice because visually aligning images with text usually involves pixel-level adjustment.

Either way, the key thing is that the behavior should match between iOS and Android so we don't need to fork the styles every time we use inline views.

**Comment by nicklockwood** [comment]
@facebook-github-bot import

**Comment by nicklockwood** [comment]
cc @dmmiller for thoughts on alignment API

**Comment by facebook-github-bot-2** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1018155508272793/int_phab) to review.

**Comment by nicklockwood** [comment]
@rigdern the alignment is still wrong. It seems like the problem is that `[layoutManager boundingRectForGlyphRange:range inTextContainer:textContainer]` doesn't return the actual frame of the NSTextAttachment, but a frame whose height matches the height of the row.

You've compensated for this by replacing the width/height with the actual width/height of the view (line 141 or RCTShadowText), but that doesn't fix the origin, which is still incorrect as it places the view at the top of the row frame instead of aligning it with the baseline.

I was able to improve this slightly by changing the logic to

      CGRect childFrame = [layoutManager boundingRectForGlyphRange:range inTextContainer:textContainer];
      childFrame = (CGRect){{
        RCTRoundPixelValue(childFrame.origin.x),
        RCTRoundPixelValue(childFrame.origin.y + childFrame.size.height - height)
      }, {
        RCTRoundPixelValue(width),
        RCTRoundPixelValue(height)
      }};

So that the view is aligned with the *bottom* of the frame instead of the top, but that's still not the same as the baseline, which is what it should be. There's presumably some way to get the actual NSTextAttachment frame instead of the frame for the placeholder glyph, but I couldn't find it after a few minutes of searching.

**Comment by facebook-github-bot-1** [comment]
@rigdern updated the pull request.

**Comment by rigdern** [comment]
@nicklockwood Sorry about that. The vertical alignment should be good now. I tested it with short and tall images and compared the results with the original implementation.

Here's what the UIExplorer example now looks like:

image

I did this as a separate commit so that you can see exactly what I changed. Let me know if you need me to squash it. In general, does your team prefer pull request revisions to be done as separate commits or squashed into the original commit?

**Comment by rigdern** [comment]
@nicklockwood Is there anything else blocking this change from being merged?

**Comment by nicklockwood** [comment]
@rigdern sorry for the delay. This looks good, thanks!

@facebook-github-bot import

**Comment by facebook-github-bot-7** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1018155508272793/int_phab) to review.

**Comment by nicklockwood** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-2** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1018155508272793/int_phab) to review.

**Comment by javache** [comment]
This regressed one of our application startup metrics by 5%. Any clue where this could be coming from? Otherwise, I'd prefer to revert this for now, until we get this metrics-neutral.

**Comment by rigdern** [comment]
@javache Without doing some analysis, I can't tell which part of this change slowed down your startup metrics. Just reverting this change fixes your startup metrics? Does your app use a lot of inline images?

@nicklockwood, do you have any thoughts on this?

**Comment by nicklockwood** [comment]
@rigdern unfortunately I haven't been able to work out what's causing the regression. I'm going to revert this for now and try landing it again next week after our branch cut.

I did find a few obvious inefficiencies, such as not stripping the views associated with RawText (these are regular UIViews, so your check for RCTText views doesn't strip them), but I tried fixing that and it had no measurable effect.

I'm guessing that the problem must be related to the extra logic added to the layout, but I don't have any ideas beyond that right now.

**Comment by rigdern** [comment]
@nicklockwood Can you provide any details about the scenario such as how many inline images there are and how the perf measurements are being taken (e.g. are you using some markers built into RN or does the app do the instrumentation itself? If the app is doing its own instrumentation, what specific timeframe is being measured and what is being used to take the measurements?)

**Comment by rigdern** [comment]
@nicklockwood I see that this got checked in again. Were you able to fix the performance problem? If so, what was the cause?

**Comment by nicklockwood** [comment]
@rigdern yes. I made some changes in RCTText which seemed to eliminate the regression, I'll annotate them for you in the commit.

**Comment by rigdern** [comment]
@nicklockwood Thanks for the explanations!

**Pull request closed without merging**